### DISCUSSION TITLE
Integral of flux over the surface is constant as BC

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#24757] [Q&A General]
Hello MOOSE experts,

I am trying to apply the NeumannBC over the surface of polycrystalline particle for diffusion problem. For the isotropic diffusion, everything is working fine as each node on the surface of the particle is feeling same flux and also the integral of J.n over the particle surface is constant. 

However, for the anisotropic case, still using the NeumannBC is giving me the negative concentration over some region. It is because, 
 the individual node on the surface of the particle should feel different flux yet the integral of J.n over the surface should be constant. Any suggestions to implement such boundary condition?

Best,
Avtar

**Comment by hugary1995** [comment]
Your understanding of the issue is correct. You'll want to use a Constraint instead of a BC. I dont think a constraint like that has already been implemented. You could follow some of the other constraints and implement one on your own.

**Reply by avtarsinghh1991** [reply]
Thanks @hugary1995. I will try to write a subroutine and will get back to you for making sure it is right as I am fairly new to C++. If you have any suggestion on which constraint subroutine I should use as reference, kindly let me know.

Best,
Avtar

**Reply by avtarsinghh1991** [reply]
Hello @hugary1995 and @GiudGiud 

For the above problem, can I use  `AverageValueConstraint` for which the postprocessor input can be from `SideDiffusiveFluxIntegral`?

Links to Average Value Constraint and Side Diffusive Flux Integral:
https://mooseframework.inl.gov/source/scalarkernels/AverageValueConstraint.html
https://mooseframework.inl.gov/source/postprocessors/SideDiffusiveFluxIntegral.html

Thanks,
Avtar

**Reply by GiudGiud** [reply]
You can try.
Imposing constraints usually makes the solve more difficult numerically so that s one drawback

**Reply by avtarsinghh1991** [reply]
Thanks @GiudGiud. 

Is there any other way to tackle the problem?

I am trying to write a subroutine for it as follows:

```
#include "IntegratedDiffusionBC.h"
 
 registerMooseObject("MooseApp",IntegratedDiffusionBC);
 
 InputParameters
 IntegratedDiffusionBC::validParams()
 {
   InputParameters params = IntegratedBC::validParams();
   params.addClassDescription("This class implements a form of the Neumann boundary condition in "
                              "which the boundary term is treated 'implicitly'.");
   params.addRequiredParam<Real>(
          "value", "Given (constant) which we want the integral of the solution variable to match.");
   return params;
 }
 
    
 IntegratedDiffusionBC::IntegratedDiffusionBC(const InputParameters & parameters) : IntegratedBC(parameters),
	    _Diff_Matrix(getMaterialProperty<RankTwoTensor>("Diff_Matrix")),
	    _value(getParam<Real>("value"))
 {
 }
 
 Real
 IntegratedDiffusionBC::computeQpResidual()
 {
   return _Diff_Matrix[_qp]*_grad_u[_qp] * _normals[_qp] * _test[_i][_qp] - _value;
 }
 
 Real
 IntegratedDiffusionBC::computeQpJacobian()
 {
   return _Diff_Matrix[_qp]*(_grad_phi[_j][_qp] * _normals[_qp]) * _test[_i][_qp];
 }
 
 Real
 IntegratedDiffusionBC::computeQpOffDiagJacobian(unsigned /*jvar*/)
 {
   // off-diagonal derivatives are all zero.
   return 0.;
 }
```

Any suggestions?

Thanks,
Avtar
